corgi/tasks/brew.py

```diff
@@ -488,34 +494,55 @@ def save_node(
     return node
 
 
-@app.task(
-    base=Singleton,
-    autoretry_for=RETRYABLE_ERRORS,
-    retry_kwargs=RETRY_KWARGS,
-    soft_time_limit=settings.CELERY_LONGEST_SOFT_TIME_LIMIT,
-)
-def load_brew_tags() -> None:
+def load_stream_brew_tags() -> None:
     for ps in ProductStream.objects.get_queryset():
-        build_type = BUILD_TYPE
-        brew = Brew(BUILD_TYPE)
-        refresh_task = slow_fetch_modular_build
-        # This should really be a property in Product Definitions
-        if settings.COMMUNITY_MODE_ENABLED and ps.name == "openstack-rdo":
-            brew = Brew(SoftwareBuild.Type.CENTOS)
-            build_type = SoftwareBuild.Type.CENTOS
-            refresh_task = slow_fetch_brew_build
+        brew, build_type, refresh_task = _relation_context_for_stream(ps.name)
         for brew_tag, inherit in ps.brew_tags.items():
-            # Always load all builds in tag when saving relations
             builds = brew.get_builds_with_tag(brew_tag, inherit=inherit, latest=False)
-            no_of_created = create_relations(
-                builds,
-                build_type,
-                brew_tag,
-                ps.name,
-                ProductComponentRelation.Type.BREW_TAG,
-                refresh_task,
-            )
-            logger.info("Saving %s new builds for %s", no_of_created, brew_tag)
+            _create_relations_for_tag(builds, brew_tag, build_type, ps.name, refresh_task)
+
+
+def load_brew_tags(build_id: str, brew_tags: list[str]) -> int:
+    all_stream_tags = ProductStream.objects.exclude(brew_tags__exact={}).values_list(
+        "name", "brew_tags"
+    )
+    no_created = 0
+    for stream_name, stream_tags in all_stream_tags:
+        for tag in brew_tags:
+            if tag in stream_tags:
+                brew, build_type, _ = _relation_context_for_stream(stream_name)
+                logger.info(f"Creating relations for {stream_name} and {tag}")
+                no_created += _create_relations_for_tag(
+                    (build_id,), tag, build_type, stream_name, None
+                )
```

Comment:
Instead of many nested loops that check all the tags, I think it's more efficient to convert the tags to a set and check the intersection:
```suggestion
    brew_tags = set(brew_tags)
    for stream_name, stream_tags in all_stream_tags:
        build_stream_tags = set(stream_tags)
        build_stream_tags = brew_tags.intersection(build_stream_tags)
        for tag in build_stream_tags:
                brew, build_type, _ = _relation_context_for_stream(stream_name)
                logger.info(f"Creating relations for {stream_name} and {tag}")
                no_created += _create_relations_for_tag(
                    (build_id,), tag, build_type, stream_name, None
                )
```

Reply:
Good idea. I'll change it.